### PR TITLE
MC-85: Allows to send categories and manage errors

### PR DESCRIPTION
**Pull request by willy-ahva**
| Q                    | A
| -------------------- | ---
| Bug fix?             | n
| New feature?         | y
| Specs pass?*         | y
| Checkstyle issues?** | n
| PMD issues?***       | n
| Fixed tickets        | MC-85
| Doc PR               |

### CODE REVIEW

File: Writer/CategoryWriter.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        $categories = [];
+        foreach ($items as $item) {
+            $categories = array_merge($categories, $item);
+        }
+
+        $mappedNames = $this->getMappedNames($categories);
+        try {
+            $this->client->exportCategories($categories);
+        } catch (\SoapFault $e) {
+            $failedCategories = json_decode($e->getMessage(), true);
```
Comment (BitOne) [inline Writer/CategoryWriter.php]: That could be another soapfault (for example denied access, or another stuff). You should check that the SoapFault code you're getting is the one matching the error you got on Magento.

File: Writer/CategoryWriter.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+ */
+class CategoryWriter extends AbstractWriter
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function write(array $items)
+    {
+        $categories = [];
+        foreach ($items as $item) {
```
Comment (BitOne) [inline Writer/CategoryWriter.php]: Hum, not really clear what you're doing here. Maybe replace those lines to put them to a function with a meaningfull name and a nice phpDoc description explaining why you're doing that.

File: Writer/AbstractWriter.php [hunk 3]
```diff
@@ -133,7 +133,7 @@ public function setEventDispatcher(EventDispatcherInterface $eventDispatcher)
      * @param array  $messageParameters
      * @param mixed  $item
      */
-    protected function addWarning($message, array $messageParameters = [], $item = null)
+    protected function addWarning($message, array $messageParameters = [], $item = [])
```
Comment (fitn) [inline Writer/AbstractWriter.php]: Why?

File: Writer/CategoryWriter.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            $categories = array_merge($categories, $item);
+        }
+
+        $mappedNames = $this->getMappedNames($categories);
```
Comment (BitOne) [inline Writer/CategoryWriter.php]: The mappedNames is only for the root category. So it should be only "mappedRoot", not names, right ?

File: Writer/CategoryWriter.php [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                $errors = $this->getFailedCategories($failedCategories, $mappedNames);
+                $this->manageFailedCategories($errors);
+            } else {
+                $this->addWarning($e->getMessage());
+            }
+        }
+    }
+
+    /**
+     * Gives lines mapped to categories name
```
Comment (BitOne) [inline Writer/CategoryWriter.php]: Hum, not clear. What are "lines" ?

File: Writer/CategoryWriter.php [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+     *
+     * @param array $categories
+     *
+     * @return array
+     */
+    protected function getMappedNames(array $categories)
```
Comment (BitOne) [inline Writer/CategoryWriter.php]: Not clear what this code is doing.

File: Writer/CategoryWriter.php [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function write(array $items)
+    {
+        $categories = [];
+        foreach ($items as $item) {
+            $categories = array_merge($categories, $item);
+        }
```
Comment (fitn) [inline Writer/CategoryWriter.php]: It does not override anything important?

File: Writer/CategoryWriter.php [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    protected function getMappedNames(array $categories)
+    {
+        $mappedNames  = [];
+        $previousName = '';
+        foreach ($categories as $key => $category) {
+            if (!empty($category[CategoryLabelDictionary::NAME_HEADER])) {
+                $mappedNames[$key] = $category[CategoryLabelDictionary::NAME_HEADER];
+                $previousName      = $category[CategoryLabelDictionary::NAME_HEADER];
+            } else {
+                $mappedNames[$key] = $previousName;
```
Comment (BitOne) [inline Writer/CategoryWriter.php]: Why do we need to put the previousName in the mappedNames of the current category ?

File: Writer/AbstractWriter.php (hunk 3 above)
Comment (willy-ahva) [inline Writer/AbstractWriter.php]: Because if I send null in stepExecution->addWarning() it breaks. It  must be of the type array.

**Pull request merged**